src/Language/PureScript/TypeChecker.hs

```diff
@@ -434,31 +434,116 @@ typeCheckModule (Module ss coms mn decls (Just exps)) =
   warnAndRethrow (addHint (ErrorInModule mn)) $ do
     modify (\s -> s { checkCurrentModule = Just mn })
     decls' <- typeCheckAll mn exps decls
+    classesToSuperClasses <- gets
+      ( M.map
+        ( S.fromList
+        . fmap constraintClass
+        . typeClassSuperclasses
+        )
+      . typeClasses
+      . checkEnv
+      )
+    let
+      -- A mapping of class names to their declaration refs in this module.
+      moduleClassDeclarations :: M.Map (Qualified (ProperName 'ClassName)) DeclarationRef
+      moduleClassDeclarations =
+        M.fromList
+          . mapMaybe (\x -> case x of
+                td@(TypeClassDeclaration _ name _ _ _ _) ->
+                  (,) (Qualified (Just mn) name) <$> classDecToRef td
+                _ ->
+                  Nothing)
+          $ decls
+      classDecToRef :: Declaration -> Maybe DeclarationRef
+      classDecToRef (TypeClassDeclaration (ss', _) n _ _ _ _) = Just (TypeClassRef ss' n)
+      classDecToRef _ = Nothing
+
+      moduleClassDeclarationSet :: S.Set (Qualified (ProperName 'ClassName))
+      moduleClassDeclarationSet = M.keysSet moduleClassDeclarations
+
+      -- A function that, given a class name, returns the set of
+      -- transitive class dependencies that are defined in this
+      -- module.
+      transitiveSuperClassesFor
+          :: Qualified (ProperName 'ClassName)
+          -> S.Set (Qualified (ProperName 'ClassName))
+      transitiveSuperClassesFor qname =
+        moduleClassDeclarationSet
+          `S.intersection`
+            untilSame
+            (\s -> s <> foldMap (\n -> fromMaybe S.empty (M.lookup n classesToSuperClasses)) s)
+            (fromMaybe S.empty (M.lookup qname classesToSuperClasses))
+
+      superClassesFor qname =
+        moduleClassDeclarationSet 
+          `S.intersection` 
+            fromMaybe S.empty (M.lookup qname classesToSuperClasses)
+
     for_ exps $ \e -> do
       checkTypesAreExported e
       checkClassMembersAreExported e
       checkClassesAreExported e
+      checkSuperClassExport superClassesFor moduleClassDeclarations e
+        `catchError` annotateSuperClasses transitiveSuperClassesFor moduleClassDeclarations
     return $ Module ss coms mn decls' (Just exps)
   where
+  qualify' :: a -> Qualified a
+  qualify' = Qualified (Just mn)
+
+  annotateSuperClasses 
```

Comment:
Could you please add a comment summarising why we only obtain transitive superclasses after the immediate check has failed?